### PR TITLE
Replace k8s.io/apimachinery/pkg/util/clock with k8s.io/utils/clock and bump k8s.io to 1.28

### PR DESCRIPTION
**Pull request by MartinForReal**
+ Replace k8s.io/apimachinery/pkg/util/clock with k8s.io/utils/clock
+ bump k8s.io to 1.28

### CODE REVIEW

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-npd-e2e-node
--


<br class="Apple-interchange-newline">
</body>
</html>

**Comment by MartinForReal** [comment]
e2e error is related to https://github.com/kubernetes/kubernetes/pull/120231/files

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-npd-e2e-node
--


<br class="Apple-interchange-newline">
</body>
</html>

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-npd-e2e-node
--


<br class="Apple-interchange-newline">
</body>
</html>

File: pkg/exporters/k8sexporter/condition/manager_test.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ import (
 	problemutil "k8s.io/node-problem-detector/pkg/util"
 
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/util/clock"
+	clock "k8s.io/utils/clock/testing"
```
Comment (hakman) [inline pkg/exporters/k8sexporter/condition/manager_test.go:32]: Should be consistent across project.
```suggestion
	testclock "k8s.io/utils/clock/testing"
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-npd-e2e-node
--


<br class="Apple-interchange-newline">
</body>
</html>

**Comment by MartinForReal** [comment]
@hakman Thanks for the fix! I've also updated the pr. PTAL. Thanks!

**Review by hakman — CHANGES_REQUESTED** [review]
@MartinForReal I don't mind picking up the changes in this PR, but maybe keep the commit as cherry-pick (not very important though).
The commit message mentions *k7s*, which is a typo. In general, this is just a dependencies upgrade and maybe a separate commit to deal with the ticker change.
I don't think the changes you added for RealClock and FakeClock are needed:
https://github.com/kubernetes/utils/blob/3b25d923346b3814e0898684c97390092f31a61e/clock/clock.go#L82C1-L83C1
```go
var _ = WithTicker(RealClock{})

// RealClock really calls time.Now()
type RealClock struct{}

...
```

**Comment by MartinForReal** [comment]
@hakman Thanks for reviewing the pr!

File: pkg/exporters/k8sexporter/k8s_exporter.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func NewExporterOrDie(ctx context.Context, npdo *options.NodeProblemDetectorOpti
 
 	ke := k8sExporter{
 		client:           c,
-		conditionManager: condition.NewConditionManager(c, clock.RealClock{}, npdo.K8sExporterHeartbeatPeriod),
+		conditionManager: condition.NewConditionManager(c, clock.WithTicker(clock.RealClock{}), npdo.K8sExporterHeartbeatPeriod),
```
Comment (hakman) [inline pkg/exporters/k8sexporter/k8s_exporter.go:62]: I don't think this change is needed.
```suggestion
		conditionManager: condition.NewConditionManager(c, clock.WithTicker(clock.RealClock{}), npdo.K8sExporterHeartbeatPeriod),
```

Comment (MartinForReal) [reply pkg/exporters/k8sexporter/k8s_exporter.go]: yes. Thanks!

File: go.mod [hunk 3]
```diff
@@ -17,22 +17,23 @@ require (
...
-	golang.org/x/sys v0.9.0
+	golang.org/x/sys v0.10.0
 	google.golang.org/api v0.114.0
-	k8s.io/api v0.17.17
-	k8s.io/apimachinery v0.17.17
-	k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible
-	k8s.io/component-base v0.17.2
+	k8s.io/api v0.28.1
+	k8s.io/apimachinery v0.28.1
+	k8s.io/client-go v9.0.0+incompatible
```
Comment (hakman) [inline go.mod:33]: Shouldn't this also `v0.28.1`?
```suggestion
	k8s.io/client-go v0.28.1
```

**Review by hakman — CHANGES_REQUESTED** [review]

File: go.mod (hunk 3 above)
Comment (MartinForReal) [reply go.mod:33]: This is one is replaced with v0.28.1. please see the last line. It is related to cadvisor. I will try to upgrade that part in following pr.

Comment (hakman) [reply go.mod:33]: Ah, ok. Thanks for the explanation.

Comment (hakman) [reply go.mod:33]: Pretty strange to depend on cadvisor just for a log tailing lib 😄

Comment (MartinForReal) [reply go.mod:33]: It has not changed for a long time..

File: pkg/util/metrics/metric_float64.go [hunk 4]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (mmiranda96) [inline pkg/util/metrics/metric_float64.go:8]: Please revert this indentation change.

File: pkg/util/metrics/metric_int64.go [hunk 5]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (mmiranda96) [inline pkg/util/metrics/metric_int64.go:8]: Please revert this indentation change.

**Review by mmiranda96 — APPROVED** [review]
Sorry for the delay! Some small comments, but overall good. Thanks!
/lgtm

File: pkg/systemlogmonitor/logwatchers/filelog/translator.go [hunk 6]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
[the remark this answers is not in the document]
Comment (hakman) [reply pkg/systemlogmonitor/logwatchers/filelog/translator.go:8]: @mmiranda96 This looks to me like a change introduced by newer `gofmt`. Could you check?

[the remark this answers is not in the document]
Comment (mmiranda96) [reply pkg/systemlogmonitor/logwatchers/filelog/translator.go:8]: You're right, my mistake.

**Comment by mmiranda96** [comment]
Please ignore my previous comments, everything looks great!

File: pkg/systemlogmonitor/logwatchers/filelog/translator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (hakman) [reply pkg/systemlogmonitor/logwatchers/filelog/translator.go:8]: Maybe set the comments as resolved, to make it easier for the final review.

**Comment by vteratipally** [comment]
sorry for the delay, 

/lgtm
/approve

**Pull request merged**